### PR TITLE
fuse, mount: Add sfsignoreflush option

### PR DESCRIPTION
**Pull request by dmga44**
The sfsignoreflush option targets answering inmediately to
the flush or fsync syscall without properly checking its responses. The
core of implementation of both syscall is write_data_flush function,
this is why is the mocked one. This fools single-file copiers into
believing the files are already copied, and the system can handle those
copies in a multithreaded fashion.

The tests performed in Windows show that to achieve the desired behavior
similar to Linux's ignoreflush, the read syscall should be also mocked
in this special mode. This is also the reason behind the renaming of
this option.

On both platforms, it has been proven the value of this option in the copy of small files.

### CODE REVIEW

**Comment by uristdwarf** [comment]
Seems fine (@lgsilva3087 should have a look before merging, I'm not entirely familiar with the Windows client), but since the CI doesn't work right now, we need to wait a bit.

In the meantime, you might want to copy edit the PR message a bit, since it's going to be squashed and included as the commit message. Someone else who's going to merge it can do it later though if you don't have time (since I can't merge it next week due to being on vacation).

File: src/mount/fuse/mount_config.cc
Comment (Antuan-AE) [inline src/mount/fuse/mount_config.cc:93]: Consider renaming the variable to match naming conventions. If the DANGEROUS tag for this option is to be highlighted, consider doing it in the documentation (help command, man pages, etc).

File: src/mount/fuse/mount_config.cc
Comment (Antuan-AE) [inline src/mount/fuse/mount_config.cc:228]: See my previous suggestion about tag DANGEROUS for naming the variable.

File: src/mount/sauna_client.h
Comment (Antuan-AE) [inline src/mount/sauna_client.h:85]: Consider using the same variable name for Unix and Windows platforms as they are intended for the same purpose and have identical initialization values.

**Review by Antuan-AE — DISMISSED** [review]
Please see my suggestions. LGTM

File: src/mount/sauna_client.cc
Comment (dmga44) [inline src/mount/sauna_client.cc:2139]: This memory is not being freed (issue notice by @aNeutrino ).

File: src/mount/sauna_client.h
Comment (dmga44) [reply src/mount/sauna_client.h:85]: Well, they are used for the equivalent options: an option which tries to fool file-by-file copiers into believing file are fast copied to make them send the request to write another file, and we could process those writes in parallel.

But the option worked differently for both OSs, for the Linux only had to mock the flush syscall, but under Windows had to mock also the read. So, I'll leave it this way because I don't want to mix both.

File: src/mount/fuse/mount_config.cc
Comment (dmga44) [reply src/mount/fuse/mount_config.cc:93]: You point out to something here, which I have just checked. It seems we don't have updated our man pages, at least the readworkers and maxreadaheadrequests are not present there. I must add them also.

I'll rename it to sfsignoreflush, but as a first word in the help and man page comment I'll the word DANGEROUS. I think it is very important that people know they may lose data when using this option.

File: src/mount/writedata.cc
Comment (rolysr) [inline src/mount/writedata.cc:216]: Consider to set this to false as it is done in some other static bool variables around the repo.

**Review by rolysr — COMMENTED** [review]
Can you consider if it would be necessary to add a test for this option?

**Review by uristdwarf — APPROVED** [review]

File: doc/sfsmount.1.adoc
Comment (lgsilva3087) [inline doc/sfsmount.1.adoc:208]: It is strange to see DANGEROUS here. Probably we can change it to:
Advanced: use with caution.

File: src/mount/fuse/mount_config.cc
Comment (lgsilva3087) [inline src/mount/fuse/mount_config.cc:227]: Same here.

File: src/mount/fuse/mount_config.cc
Comment (lgsilva3087) [inline src/mount/fuse/mount_config.cc:94]: Is it enabled by default or disabled?

**Review by lgsilva3087 — CHANGES_REQUESTED** [review]
It is strange to have similar options with different names in Linux and Windows (ignore_flush and special_write_mode + ignore_read). I also don't like to ignore the reads, which makes the mount point write-only in Windows.
From your comments, looks like the Windows copier is triggering the read, if it is the case, we should not ignore the reads and just recommend another copier for small files.

File: src/mount/fuse/mount_config.cc
Comment (dmga44) [reply src/mount/fuse/mount_config.cc:94]: It is disabled by default.

**Comment by dmga44** [comment]
> just recommend another copier for small files.

Well, I've already performed extra tests and checked that Ultracopier, FastCopy and our custom sfscp also perform the extra read operation. Looks like it is even tied to the ```std::filesystem::copy``` which we are using in the sfscp tool. I was expecting some positive result from some of them, but it seems that the copier should be based on much lower level operations, like directs open, write, close.

File: src/mount/fuse/mount_config.cc
Comment (aNeutrino) [reply src/mount/fuse/mount_config.cc:227]: yes lets remove DANGEROUS here....

File: doc/sfsmount.1.adoc
Comment (aNeutrino) [reply doc/sfsmount.1.adoc:208]: maybe we can reuse the same text without DANGEROUS... from below:

Ignore flush usual behavior by replying SUCCESS to it immediately. Targets fast creation of small files, but may cause data loss during crashes."

File: src/mount/fuse/mount_config.cc
Comment (aNeutrino) [reply src/mount/fuse/mount_config.cc:227]: @dmga44 typo in word: ~inmediately~ - should be: immediately

File: doc/sfsmount.1.adoc
Comment (dmga44) [reply doc/sfsmount.1.adoc]: Done.

File: src/mount/fuse/mount_config.cc
Comment (dmga44) [reply src/mount/fuse/mount_config.cc]: Done, both.

**Review by lgsilva3087 — APPROVED** [review]
LGTM.

**Comment by lgsilva3087** [comment]
> LGTM.

Please, squash before merge.

**Pull request merged**
